src/rules/eva-voice-biometric.js

```diff
@@ -0,0 +1,165 @@
+ *	EVA WEB is an offering from Auraya Systems.
+ *
+ *	The voice tokens used with EVA WEB are predominantly digit strings, although phrases are also supported.
+ *	The usual verification process will usually prompt the user to read out a sequence of random digits,
+ *	a sequence of consecutive digits, and optionally their phone number.
+ *
+ *	The first time a user is directed to EVA WEB where they will be invited to enrol their voice.
+ *	This process captures between 3 and 8 utterances depending on the required voice tokens, and includes a biometric “Opt-In” consent step.
+ *	At the conclusion of enrolment they will be redirected back to the authentication flow.
+ *
+ *	Configuration Items:
+ *	AURAYA_URL = EVA endpoint, typically: https://eva-web.mydomain.com/server/oauth
+ *	AURAYA_CLIENT_ID = JWT client id on the EVA server (and this server)
+ *	AURAYA_CLIENT_SECRET = JWT client secret on the EVA server (and this server)
+ *	AURAYA_ISSUER = this app (or "issuer")
+ *
+ *	AURAYA_RANDOM_DIGITS = true|false whether to prompt for random digits
+ *	AURAYA_COMMON_DIGITS = true|false whether to prompt for common digits
+ *	AURAYA_PERSONAL_DIGITS = a user.user_metadata field that contains digits such as phone_number, or blank
+ *
+ *	AURAYA_DEBUG = true|false (or blank) controls debug output
+ */
+function evaVoiceBiometric(user, context, callback) {
+  if (
+    typeof configuration.AURAYA_DEBUG !== 'undefined' &&
+    configuration.AURAYA_DEBUG === 'true'
+  ) {
+    console.log(user);
+    console.log(context);
+    console.log(configuration);
+  }
+
+  if (
+    !configuration.AURAYA_COMMON_DIGITS ||
+    !configuration.AURAYA_RANDOM_DIGITS ||
+    !configuration.AURAYA_PERSONAL_DIGITS ||
+    !configuration.AURAYA_URL ||
+    !configuration.AURAYA_CLIENT_ID ||
+    !configuration.AURAYA_CLIENT_SECRET ||
+    !configuration.AURAYA_ISSUER
+  ) {
+    console.log('Missing required configuration. Skipping.');
+    return callback(null, user, context);
+  }
+
+  const clientSecret = configuration.AURAYA_CLIENT_SECRET;
+
+  // Prepare user's enrolment status
+  user.user_metadata = user.user_metadata || {};
+  user.user_metadata.auraya_eva = user.user_metadata.auraya_eva || {};
+
+  // User has initiated a login and is prompted to use voice biometrics
+  // Send user's information and query params in a JWT to avoid tampering
+  function createToken(user) {
+    const options = {
+      expiresInMinutes: 2,
+      audience: configuration.AURAYA_CLIENT_ID,
+      issuer: configuration.AURAYA_ISSUER
+    };
+
+    return jwt.sign(user, clientSecret, options);
+  }
+
+  if (context.protocol === 'redirect-callback') {
+    // User was redirected to the /continue endpoint with correct state parameter value
+
+    var options = {
+      //subject: user.user_id, // validating the subject is nice to have but not strictly necessary
+      jwtid: user.jti
```

Comment:
I added a comment about this `jti` value: my understanding is that unlike `state`, this value can't be spoofed by DNS hacking or inspecting the payload since you need the secret key to decode it.